src/envoy/auth/jwt.cc

```diff
@@ -0,0 +1,197 @@
+/* Copyright 2017 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jwt.h"
+
+#include "common/common/base64.h"
+#include "common/common/utility.h"
+#include "openssl/evp.h"
+#include "openssl/rsa.h"
+#include "rapidjson/document.h"
+
+#include <algorithm>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace Envoy {
+namespace Http {
+namespace Auth {
+
+namespace Base64url {
+
+std::string decode(std::string input) {
+  /*
+   * base64url is using '-', '_' instead of '+', '/' in base64 string.
+   */
+  std::replace(input.begin(), input.end(), '-', '+');
+  std::replace(input.begin(), input.end(), '_', '/');
+  /*
+   * base64 string should be padded with '=' so as to the length of the string
+   * is divisible by 4.
+   */
+  switch (input.length() % 4) {
+    case 0:
+      break;
+    case 2:
+      input += "==";
+      break;
+    case 3:
+      input += "=";
+      break;
+    default:
+      /*
+       * for invalid base64url input, return empty string
+       */
+      return "";
+  }
+  return Base64::decode(input);
```

Comment:
I don't think it's a bug, it's just assuming the validity of input, as 4n-th or (4n+1)-th character never be a padding in a correct base64 string.
So what we need is to check the validity of input string of base64url.